### ISSUE TITLE
Should sd-bus send BEGIN at the same time as AUTH EXTERNAL?

### ISSUE THREAD

**Issue by refi64** [systemd/systemd#16610]
**systemd version the issue has been seen with**
> 245

**Used distribution**
 > This was reproduced in the Freedesktop SDK, 19.08 and 20.08 beta. Afaik no relevant config options should affect this.

**Expected behaviour you didn't see**
 
The D-Bus spec mentions for the [BEGIN command](https://dbus.freedesktop.org/doc/dbus-specification.html#auth-command-begin):

> The BEGIN command acknowledges that the client has received an OK command from the server and completed any feature negotiation that it wishes to do, and declares that the stream of messages is about to begin.

**Unexpected behaviour you saw**
 
sd-bus [sends the BEGIN command right along with AUTH EXTERNAL](https://github.com/systemd/systemd/blob/v245/src/libsystemd/sd-bus/bus-socket.c#L646-L670), rather than using it to acknowledge an OK. I found this because it confuses Flatpak's xdg-dbus-proxy, which generally expects BEGIN to be sent after the auth is complete.

After playing around with it for a while, I do think xdg-dbus-proxy is right here, simply because it seems weird to acknowledge a command you haven't received yet... :sweat_smile: 

**Steps to reproduce the problem**

Don't have a clean reproducer, I mean this is conceptual problem too anyway :sweat_smile:

**Comment by poettering** [comment]
We pipeline this to avoid roundtrips. Waiting for the roundtrip needlessly really slows things down needlessly.

Seems like a bug in xdg-dbus-proxy. It should process client requests in order, and hence not get confused.

Yes, this means if authentication fails, then we'll send the begin anyway, but that shouldn't be a problem, as the server should then simply return failure which will be handled correctly by the client.

I don't see any bug here in sd-bus. the dbus proxy in flatpak should be prepared for such pipelining, like dbus-daemon and dbus-broker are.

I am tempted to just close this.

Have you reported this to the flatpark devs?

**Comment by refi64** [comment]
Hmm I understand the idea and will look into fixing it there. Technically not a D-Bus issue, but would it be worth adding a note on this to the spec? From what I could tell it didn't mention this technique at all, which is what led to the confusion.

**Comment by poettering** [comment]
I am not sure this needs to be documented. It appears to be a timing issue only, right? i.e. xdg-dbus-proxy doesn't expect the follow-up command to appears so quickly. But timing issues like this are always bugs in network protocols: every participant has to be prepared to receive everything as slow or as quick as the other side wants and needs to make the best of it.

Anyway, consider opening an explicit request on dbus gitlab instance, if you want this documented.

@smcv might have something to say about this too?

(btw, I think sd-bus should actually be even more aggressive with the pipelining than it actually is: it currently waits for a roundtrip on the BEGIN reply. Ideally it would already enqueue the initial D-Bus Hello() message in one go too. As well as any user-provided message it might want to enqueue. That way we wouldn't pay for connection setup anything anymore, and purely wait on the dbus reply only.)

**Comment by smcv** [comment]
I'm not at all sure that the D-Bus Specification even intended to allow pipelining - the fact that it works seems more like an implementation quirk of the reference implementation. (I didn't write the spec, but to me it doesn't read as though pipelining was expected or intended.)

> btw, I think sd-bus should actually be even more aggressive with the pipelining than it actually is: it currently waits for a roundtrip on the BEGIN reply. Ideally it would already enqueue the initial D-Bus Hello() message in one go too. As well as any user-provided message it might want to enqueue

You can't do that in general: because AF_UNIX SOCK_STREAM sockets behave like a strange hybrid of datagram and stream, and because the reference dbus-daemon is not expecting pipelining, any out-of-band Unix fds sent before authentication has completed will get lost. https://gitlab.freedesktop.org/dbus/dbus/-/issues/182

**Comment by refi64** [comment]
> I am not sure this needs to be documented. It appears to be a timing issue only, right? i.e. xdg-dbus-proxy doesn't expect the follow-up command to appears so quickly. But timing issues like this are always bugs in network protocols: every participant has to be prepared to receive everything as slow or as quick as the other side wants and needs to make the best of it.

To be clear, it's because BEGIN is being sent in the same message as AUTH EXTERNAL, but xdg-dbus-proxy is expecting BEGIN to be sent after the OK response, i.e. after AUTH EXTERNAL is sent and acknowledged.

I'd argue that this really is either an sd-bus bug or something missing in the spec, just because it *explicitly says* BEGIN is acknowledging an OK, which would imply the client already received the OK.

**Comment by poettering** [comment]
> To be clear, it's because BEGIN is being sent in the same message as AUTH EXTERNAL, but xdg-dbus-proxy is expecting BEGIN to be sent after the OK response, i.e. after AUTH EXTERNAL is sent and acknowledged.

It's a stream socket. There are no "messages". The protocol implementation for any kind of stream protocol should read as much as it can process and process the rest in a later iteration. i.e. in the case of the dbus auth proto, read until newline, then process that, and process the rest later.

gdbus, dbus-daemon, dbus-broker, sd-bus all can handle this correctly, and it cuts authentication short substantially, making bus connections almost as cheap latency-wise as raw AF_UNIX sockets. I am not convinced we should fuck that up, just because flatpak can't be fixed to handle this correctly.

BEGIN acknowledges the OK that we are pretty sure to receive shortly... Inserting an artificial delay just makes things needlessly slow.

If the flatpak bus proxy wants to be considered a thin proxy around dbus-daemon or dbus-broker it better act like it, and that means behave reasonably like it, where it makes a lot of sense.

Latency matters. In UIs it's kinda the only thing that matters. Don't fuck this up just because flatpak doesn't want to fix that.

systemd uses D-Bus a lot. We are not going to insert additional synchronization points into dbus connection set-up in sd-bus, because that would slow down the whole system measurably. I am sorry. If flatpak won't fix this, we can certainly add an env var that inserts an artifical sleep(1) somewhere, but I am not sure anyone would be helped with that.

In fact, the gdbus authentication code, with it's roundtrip-happiness was one of the reasons why we ended up doing sd-bus, because these roundtrips are exactly how this should *not* look like I am sure.

**Comment by smcv** [comment]
> It's a stream socket. There are no "messages".

Yes, except no: the boundaries between sendmsg() calls are observable in some cases at the level of the kernel <-> user-space interface, in particular where out-of-band data (fds, credentials) get attached (at the kernel level I believe they're getting attached to a skbuf, although my kernel knowledge is second-hand and incomplete).

**Comment by poettering** [comment]
> Yes, except no: the boundaries between sendmsg() calls are observable in some cases at the level of the kernel <-> user-space interface, in particular where out-of-band data (fds, credentials) get attached (at the kernel level I believe they're getting attached to a skbuf, although my kernel knowledge is second-hand and incomplete).

"in particular"? I would say "only"... effectively they are just byte streams with some corner cases. And given that dbus supports tcp as backend, too, from dbus perspectively they are primarily byte streams too, except that quirks exist for some specific cases. I mean, the auth protocol itself (after the initial zero byte) doesn't bother with SCM_CREDENTIALS or SCM_RIGHTS, so it even *is* purely stream based here...

**Comment by smcv** [comment]
I've opened flatpak/xdg-dbus-proxy#21 to approach this interop failure from the other end.

**Comment by poettering** [comment]
Fixed in xdg-dbus-proxy:

https://github.com/flatpak/xdg-dbus-proxy/commit/4607d5855bea8802d1fe93f21f1e9c962320d775
